waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java

```diff
@@ -252,62 +258,110 @@ private boolean isWhitelisted(String databasePrefix, String database) {
   public PanopticOperationHandler getPanopticOperationHandler() {
     return new PanopticOperationHandler() {
 
+      private void shutdownExecutorService(ExecutorService executorService) {
+        executorService.shutdown();
+        try {
+          if (!executorService.awaitTermination(600, TimeUnit.MILLISECONDS)) {
+            executorService.shutdownNow();
+          }
+        } catch (InterruptedException e) {
+          executorService.shutdownNow();
+        }
+      }
+
+      private void getResultFromFutures(List<Future<?>> futures) {
+        for (Future<?> future : futures) {
+          try {
+            future.get(200, TimeUnit.MILLISECONDS);
+          } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            LOG.warn("Can't fetch databases by pattern: {}", e.getMessage());
+          }
+        }
+      }
+
       @Override
       public List<TableMeta> getTableMeta(String db_patterns, String tbl_patterns, List<String> tbl_types) {
         List<TableMeta> combined = new ArrayList<>();
-        for (Entry<DatabaseMapping, String> mappingWithPattern : databaseMappingsByDbPattern(db_patterns).entrySet()) {
-          try {
-            DatabaseMapping mapping = mappingWithPattern.getKey();
-            String patterns = mappingWithPattern.getValue();
-            List<TableMeta> tables = mapping.getClient().get_table_meta(patterns, tbl_patterns, tbl_types);
-            for (TableMeta tableMeta : tables) {
-              if (isWhitelisted(mapping.getDatabasePrefix(), tableMeta.getDbName())) {
-                combined.add(mapping.transformOutboundTableMeta(tableMeta));
+        Map<DatabaseMapping, String> databaseMappingsForPattern = databaseMappingsByDbPattern(db_patterns);
+        ExecutorService executorService = Executors.newFixedThreadPool(databaseMappingsForPattern.size());
+        List<Future<?>> futures = new ArrayList<Future<?>>();
+
+        for (Entry<DatabaseMapping, String> mappingWithPattern : databaseMappingsForPattern.entrySet()) {
+          Runnable runnableTask = () -> {
+            try {
+              DatabaseMapping mapping = mappingWithPattern.getKey();
+              String patterns = mappingWithPattern.getValue();
+              List<TableMeta> tables = mapping.getClient().get_table_meta(patterns, tbl_patterns, tbl_types);
+              for (TableMeta tableMeta : tables) {
+                if (isWhitelisted(mapping.getDatabasePrefix(), tableMeta.getDbName())) {
+                  combined.add(mapping.transformOutboundTableMeta(tableMeta));
```

Comment:
Could different threads adding to a shared list could lead to some issues? For example if one thread resizes the list while another is inserting. Maybe it would be better to use `Callable` and collect the output from each thread at the end, or perhaps there is a thread safe list implementation which can be used.

Reply:
Just saw Patrick's suggestion, apologies for the duplicate

Reply:
Just changed `Runnable` to `Callable` hoping to preserve the order of elements added in the `combined` list (although it still doesn't fully work yet 🤔)